keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -526,302 +517,382 @@ dualstack cluster can have this flag configured as
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
+  preferDualStack: true # set by user
+  ipFamilies:           # defaulted
+    - IPv6              # note single entry
+  clusterIP: 2001::1    # allocated
+  clusterIPs:           # allocated
+    - 2001::1           # note single entry
+```
+
+On a dual stack cluster it would have produced:
 
-# dual stack service (optional)
+```yaml
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
   type: ClusterIP
-  preferDualStack: true # set by user
-  ipFamilies:
-    - IPv6 # set by user
-    - IPv4 # defaulted (cluster supports dual stack)
-  clusterIP: 2001::1  # allocated
-  clusterIPs: # allocated
-    - 2001::1
-    - 1.2.3.4
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
-
+  preferDualStack: true # set by user
+  ipFamilies:           # defaulted
+    - IPv6              # note two entries
+    - IPv4 
+  clusterIP: 2001::1    # allocated
+  clusterIPs:           # allocated
+    - 2001::1           # note two entries
+    - 1.2.3.4
 ```
 
 ##### Notes on NodePort Allocation
 
-NodePort reservation will be constant across both families. That mean a
-reservation of `NodePort: 12345` will happen across both families irrespective
-of the IP family of the service. `kube-proxy` will ensure that traffic is
+NodePort reservations will apply across both families. That means a
+reservation of `NodePort: 12345` will happen on both families, irrespective
+of the IP family of the service. A single service which is dual stack may use
+the same NodePort on both families.  `kube-proxy` will ensure that traffic is
 routed according to the families assigned for services.
 
 #### Type Headless services
 
-1. For Headless with selector, the value of `spec.preferDualStack` and
-   `spec.ipFamilies` will drive the endpoint selection. Endpoint slice
-   controller will select endpoints with IP families that match the value of
-   these two flags.
+1. For Headless with selector, the value of `spec.ipFamilies` will drive the
+   endpoint selection. Endpoints will use the value of `ipFamiles[0]`.
+   EndpointSlices will use endpoints with all IP families of the Service.
 2. For Headless without selector the values of `spec.preferDualStack` and
-   `spec.ipFamilies` are treated as saved as they were provided by the user.
+   `spec.ipFamilies` are saved as they were provided by the user.
 
-> in both cases the values of `spec.preferDualStack` and `spec.ipFamilies` are
-> validated semantically.
+> In all cases, the values of `spec.preferDualStack`, `spec.ipFamilies`, and
+> `spec.clusterIPs` are validated and must be consistent.
 
 #### Type ExternalName
 
-For ExternalName service the value of `spec.ipFamilyPolicy` and
+For ExternalName service the value of `spec.preferDualStack` is ignored and
 `spec.ipFamilies` is expected to be `nil`. Values will be set to `nil` by the
```

Comment:
"expected" is vague. Does that mean it's a validation failure if it's not?

Reply:
@khenidak ?

Reply:
validation error, the new PR specifically calls this out